### ISSUE TITLE
Captcha_OCR How do I solve this function to return the right value? 

### ISSUE THREAD

**Issue by gabriel-fsa** [keras-team/keras-io#327]
I train the network ([here](https://keras.io/examples/vision/captcha_ocr/)) and save using the (save) method, then load the training network (load_model) and I created this function to pass the photo and return the value, but it does not come with correct values.

`
 def broken_captcha(image_path, label):

	model = keras.models.load_model('./redeTreinada')
 
	model.summary()

	validation_dataset = tf.data.Dataset.from_tensor_slices((np.array([image_path]), np.array([label])))
	validation_dataset = (
		validation_dataset.map(
			encode_single_sample, num_parallel_calls=tf.data.experimental.AUTOTUNE
		)
		.batch(batch_size)
		.prefetch(buffer_size=tf.data.experimental.AUTOTUNE)
	)
 
	prediction_model = keras.models.Model(
		model.get_layer(name="image").input, model.get_layer(name="dense2").output
	)

	def decode_batch_predictions(pred):
		input_len = np.ones(pred.shape[0]) * pred.shape[1]
		# Use greedy search. For complex tasks, you can use beam search
		results = keras.backend.ctc_decode(pred, input_length=input_len, greedy=True)[0][0][
			:, :max_length
		]
		# Iterate over the results and get back the text
		output_text = []
		for res in results:
			res = tf.strings.reduce_join(num_to_char(res)).numpy().decode("utf-8")
			output_text.append(res)
		return output_text

	for batch in validation_dataset.take(1):
		batch_images = batch["image"]
		batch_labels = batch["label"]
		print(batch_images,'-')
		print(batch_labels,'--------')

		preds = prediction_model.predict(batch_images)
		pred_texts = decode_batch_predictions(preds)

		orig_texts = []
		for label in batch_labels:
			label = tf.strings.reduce_join(num_to_char(label)).numpy().decode("utf-8")
			orig_texts.append(label)

		_, ax = plt.subplots(4, 4, figsize=(15, 5))
		for i in range(len(pred_texts)):
			img = (batch_images[i, :, :, 0] * 255).numpy().astype(np.uint8)
			img = img.T
			title = f"Prediction: {pred_texts[i]}"
			ax[i // 4, i % 4].imshow(img, cmap="gray")
			ax[i // 4, i % 4].set_title(title)
			ax[i // 4, i % 4].axis("off")
	plt.show()
 
broken_captcha('./2b827.png', '2b827')`

And the result is wrong.
![Screenshot_1](https://user-images.githubusercontent.com/55073421/101520970-f03ce400-3963-11eb-98de-2e79001ba681.png)

Can someone help me?

**Comment by TimbusCalin** [comment]
I also ran into this phenomenon and the predictions are terrible when loading the model. I train the model, I get good results, when I try to load it regardless of load_weights or load_model, the results are horrendous.

**Comment by TimbusCalin** [comment]
I suspect it is the layer which converts the characters.... the results are not exactly random, I checked two similar ground truth labels with the exact correct predictions.

**Comment by TimbusCalin** [comment]
I was write, this is the bug dear @gabriel-fsa.

The problem is the fact that the output of the char_to_num() layer is not deterministic.

I ran the example in Jupyter Notebooks, and these are the results . I have 20 different characters. 

I ran a notebook session, and printed the mappings.

########## JUPYTER NOTEBOOK SESSION 1 MAPPINGS #########################
['P', '3', 'X', '0', '2', '9', 'L', 'U', 'T', '1', 'A', 'E', 'O', 'M', '-', '7', '6', '4', '5', '8']
tf.Tensor([ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20], shape=(20,), dtype=int64)


###############JUPYTER NOTEBOOK SESSION 2 MAPPINGS###########################


['M', '7', '3', 'L', '6', '2', 'P', 'A', '1', 'T', '-', '4', 'X', '8', '5', '9', 'O', 'E', '0', 'U']
tf.Tensor([ 1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20], shape=(20,), dtype=int64)

**Comment by TimbusCalin** [comment]
Here Gabriel,

This is what you have to do:


`characters = sorted(characters)`

In this way, we ensure an ordering of the characters so that the result is always reproducible.

**Comment by MauroLondon** [comment]
Hello TimbusCalin, you know I have 3 days investigating the same problem. I have burned my eyelashes looking for a solution and it had not occurred to me to seek help through that means. I will immediately test the proposed solution ... Thanks in advance

**Comment by TimbusCalin** [comment]
@MauroLondon  I am sure it will solve your problem, as it did in my case.

**Comment by MauroLondon** [comment]
> @MauroLondon I am sure it will solve your problem, as it did in my case.

It works beautiful, you are a crack. Thank you very much for that solution

**Comment by gabriel-fsa** [comment]
@TimbusCalin Dude, you helped a lot. I am interning and one of my activities is to create and train this neural network to become a micro service. I'm trying to adapt this neural network to solve slightly different captchas (with different sizes), and I'm having some difficulties (this is normal). Are you doing the same? We could try to develop together, with a group on whatsapp. If you're interested, send a message. And sorry for my English kkk, it's very bad.

**Comment by hei-my** [comment]
> @TimbusCalin Dude，您提供了很多帮助。我正在实习，我的活动之一是创建并训练该神经网络，使其成为微服务。我正在尝试使这个神经网络适应稍微不同的验证码（大小不同），并且遇到一些困难（这是正常的）。你也一样吗 我们可以尝试与whatsapp小组一起发展。如果您有兴趣，请发送一条消息。很抱歉，我的英语kkk非常糟糕。

`labels = [os.path.splitext(os.path.split(img)[-1])[0] for img in images]
tmp = []
for label in labels:
    if len(label) < 6:
        label += '\t'*(6-len(label))
    tmp.append(label)

labels = tmp`

**Comment by hei-my** [comment]
> @MauroLondon I am sure it will solve your problem, as it did in my case.

I used 30,000 pictures for training, in which there were three label lengths, 4, 5 and 6. 

```
images = sorted(list(map(str, list(data_dir.glob("*")))))
labels = [os.path.splitext(os.path.split(img)[-1])[0] for img in images]
tmp = []
for label in labels:
    if len(label) < 6:
        label += '\t'*(6-len(label))
    tmp.append(label)

labels = tmp
```

Then I supplemented the label length to the length of 6 uniformly, and then I also used your method sorted to fix the character order.
The final training result is `
Epoch 93/500000
2195/2195 [==============================] - 270s 123ms/step - loss: 0.2727 - val_loss: 0.3048
`.
But the model does not have the generalization ability, and it cannot predict the correct result on the picture it has never seen before. Could you please help me to see why?

**Comment by sachinprasadhs** [comment]
Hi, 

The issue seems to have fixed with the commit https://github.com/keras-team/keras-io/commit/66f43291a8ad3b8a27b391811c8f2bd58ba4fec1.
Could you please verify and close the issue. Thanks!
